### ISSUE TITLE
Potential bug in _compute_sampled_logits when `num_true` > 1

### ISSUE THREAD

**Issue by geopapa11** [tensorflow/tensorflow#32659]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): macOS Mojave (10.14.6)
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): source
- TensorFlow version (use command below): 1.14.0
- Python version: 3.6.4
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory:

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with: 1. TF 1.0: `python -c "import
tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"` 2. TF 2.0: `python -c
"import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
When logits are computed in `_compute_sampled_logits` for the evaluation of `tf.nn.nce_loss`, the output related to the true labels from `out_labels` might be wrong. 

Let me provide an example. Assume `batch_size = 2`, `num_true = 1000`, `num_sampled = 20`, then `out_labels` will be a Tensor of size `2 x 1020` (1020 is the addition of `num_true` and `num_sampled`).
According to the following line:
`array_ops.ones_like(true_logits) / num_true` (one of the last lines of `_compute_sampled_logits` function),
the first 1000 elements in each row of `out_labels` are: 1./1000 = 0.001, while the last 20 are 0.

The potential problem is that the output of `_compute_sampled_logits` is then fed to a sigmoid cross entropy (**not** softmax cross entropy which requires the labels to sum to 1 in each row). As a result, labels of an example are independent to each other. This means we can have as many labels as possible for each example, and `labels` do not necessarily have to sum to 1 (for each example).

By dividing `array_ops.ones_like(true_logits)` by `num_true`, the weights for the first 1000 (positive) labels of each example would try to produce a `y_hat` that would be close to 0.001 instead of 1. Predictions, which are generated via a sigmoid function, represent the probability that a given label is positive (for this example). Let's call this prediction, `y_hat` for simplicity. Due to the above formulation, `y_hat` would try to be as close to value `1/num_true` (instead of 1) for labels that appeared with the corresponding x feature.

**Describe the expected behavior**
The expected behavior would be that all positive labels in each example be assigned value 1 instead of `1/num_true`. This way, the sigmoid which applies independently to each label will try to produce a `y_hat` for each (positive) label close to 1 instead of `1/num_true`.

The solution would be:
`    out_labels = array_ops.concat([array_ops.ones_like(true_logits), array_ops.zeros_like(sampled_logits)], 1)
`

**Code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate the problem.

```
import tensorflow as tf
import pandas as pd
import numpy as np
import math

batch_size = 4
num_classes = 2000
dim = 16
num_sampled = 20
num_true = 1000
sampled_values = None
remove_accidental_hits = False
partition_strategy = 'div'

weights = tf.get_variable('weights', initializer=tf.truncated_normal([num_classes, dim], stddev=1.0 / math.sqrt(dim)), trainable=True)
biases = tf.get_variable('biases', initializer=tf.zeros([num_classes]), trainable=True)
inputs = tf.random_normal([batch_size, dim])
df = pd.DataFrame({'labels': [np.array([0, 2], np.int32), np.array([1], np.int32),
                              np.array([6, 4, 1, 0, 3, 2], np.int32), np.array([5, 0, 2, 1, 6, 3, 9, 4], np.int32)]})
labels = tf.keras.preprocessing.sequence.pad_sequences(df['labels'].values, num_true, dtype=np.int64, value=0)
labels = tf.constant(labels)

logits, labels = _compute_sampled_logits(
    weights=weights,
    biases=biases,
    labels=labels,
    inputs=inputs,
    num_sampled=num_sampled,
    num_classes=num_classes,
    num_true=num_true,
    sampled_values=sampled_values,
    subtract_log_q=True,
    remove_accidental_hits=remove_accidental_hits,
    partition_strategy=partition_strategy)
```

Here, `_compute_sampled_logits` is from `tensorflow.python.ops.nn_impl`.

**Comment by ravikyram** [comment]
Looks like code is incomplete. In order to expedite the trouble-shooting process, please provide a minimal standalone code to reproduce the issue reported here. Thanks!

**Comment by geopapa11** [comment]
Hi @ravikyram thank you very much for your response! Thinking about it a bit more carefully, the code might probably be correct if you output predictions via a softmax function.

To extend a little bit more on what I meant above, let's assume `inputs` is a `batch_size` x `1024` matrix, `nce_weights` is `num_classes` x `1024` and `nce_biases` is `num_classes` x `1`.

Then, `all_logits = tf.nn.bias_add(tf.matmul(inputs, nce_weights, transpose_b=True), nce_biases)`
is a `batch_size` x `num_classes` matrix. An element `all_logits[i,j]` represents the logit for label `j` (of example `i`).

For each row of `all_logits` we select `num_true` columns that correspond to the true labels of this example (row). If, for example, the first row (example) is comprised of `num_true = 3` labels, which are 1, 4, 6, we will select the 1st, 4th, 6th column from the first row of `all_logits` (we assume here that columns start indexed at zero). These are the `true_logits`. In addition, we select `num_sampled` columns from each row of `all_logits` representing the negative label (0). These are the `sampled_logits`.

`out_logits`, which is concatenation of `true_logits` and `sampled_logits`, is a `batch_size` x (`num_true` + `num_sampled`) matrix. These `out_logits` are aligned with `out_labels` as follows:
```
out_labels = [[1./num_true  1./num_true  1./num_true ... 1./num_true 0 0 ... 0],
              [1./num_true  1./num_true  1./num_true ... 1./num_true 0 0 ... 0],
               ...
              [1./num_true  1./num_true  1./num_true ... 1./num_true 0 0 ... 0]]
```
An element of `out_logits` corresponding to a true label would try to match `1./num_true` via the sigmoid: `1./(1 + exp(-out_logits)) ≈ 1./num_true`. 

Now, if we used a **sigmoid** function to retrieve predictions later on (and assuming the NCE loss did a good job bringing `out_logits` close to `out_labels`), the result of the prediction might have been deceptive. The predicted value for a label would be `1./(1 + exp(-out_logits))` and its maximum value (in places of a true label) would have been close to `1./num_true` (that is, far from 1). However, if we retrieve predictions via a **softmax** function instead, then the max value (in places of a true label) would be something close to `1./num_true`, which is ok since softmax imposes a valid distribution over the `num_classes` labels.
